### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by trifonova-n**
Added metric for performance estimation

### CODE REVIEW

File: performance/estimator.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+
+from fid.inception import InceptionV3
+from fid.fid_score import calculate_activation_statistics, calculate_frechet_distance, get_activations
+import torch
+import numpy as np
+
+
+class Estimator(object):
```
Comment (Slonegg) [inline performance/estimator.py]: I guess no point having this class yet.

File: performance/estimator.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        s1 = np.cov(act_real, rowvar=False)
+        m2 = np.mean(act_fake, axis=0)
+        s2 = np.cov(act_fake, rowvar=False)
+        fid_value = calculate_frechet_distance(m1, s1, m2, s2)
+        return fid_value
+
+    def distance(self, X1, X2):
+        assert(X1.shape == X2.shape)
+        X1 += 1
+        X1 /= 2
```
Comment (Slonegg) [inline performance/estimator.py:71]: is `X1 = (X1 + 1) / 2` bad?

File: performance/estimator.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import torch
+import numpy as np
+
+
+class Estimator(object):
+    def __init__(self):
+        pass
+
+
+class FIDEstimator(Estimator):
```
Comment (Slonegg) [inline performance/estimator.py]: Add some comments to the class definition, function definitions and important or not obvious parameters.

File: performance/estimator.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        :param generator:
+        :param loader:
+        :return:
+        """
+        generator.eval()
+        act_real = []
+        act_fake = []
+        for idx, sample in zip(range(self.limit), loader):
+            noise = self.noise_sampler.sample_batch(sample[0].shape[0])
+            noise = [c.to(self.generator_device) for c in noise]
```
Comment (Slonegg) [inline performance/estimator.py:44]: Comments like "transfer noise to the generator device (GPU)"

File: performance/estimator.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+class Estimator(object):
+    def __init__(self):
+        pass
+
+
+class FIDEstimator(Estimator):
+    def __init__(self, noise_sampler, config, limit=100):
+        super().__init__()
+        self.dims = 2048
+        block_idx = InceptionV3.BLOCK_INDEX_BY_DIM[self.dims]
```
Comment (Slonegg) [inline performance/estimator.py]: move block_idx to the line where it is actually used

File: inpainting/visualizer.py [hunk 6]
```diff
@@ -37,12 +38,18 @@ def update_losses(self, epoch, g_loss, d_loss, type):
                       opts=dict(title=type + " losses", legend=['generator', 'discriminator']))
         print("Update losses")
 
+    def update_plot(self, x, y, name):
+        Y = np.array([y])
```
Comment (Slonegg) [inline inpainting/visualizer.py:49]: somewhat weird that you need to add dimension and convert to the array` np.array([y])`

File: inpainting/dataset.py [hunk 7]
```diff
@@ -102,13 +102,13 @@ def sample(self):
...
         # we need tuple here to have interface consistent with ConditionSampler
-        return torch.from_numpy(z).cuda(),
+        return torch.from_numpy(z),
 
     def sample_batch(self, batch_size):
         z = np.random.uniform(-1., 1.0, size=(batch_size, self.z_size)).astype(np.float32)
         # tuple with 1 element
         # we need tuple here to have interface consistent with ConditionSampler
-        return torch.from_numpy(z).cuda(),
+        return torch.from_numpy(z),
```
Comment (Slonegg) [inline inpainting/dataset.py]: I would add braces like `(torch.from_numpy(z),)`

File: gan/trainer.py [hunk 8]
```diff
@@ -31,19 +33,20 @@ def __init__(self, generator, discriminator, config, noise_sampler, lr=0.0002, v
...
+        with torch.cuda.device(self.device.index):
+            try:
+                last_epoch = self.current_epoch
+                for self.current_epoch in range(last_epoch + 1, n_epochs + 1):
+                    self.train_epoch(train_loader)
+                    self.save_checkpoint()
+                    if valid_loader is not None:
+                        self.valid_epoch(valid_loader)
+            except Exception as e:
+                import traceback
```
Comment (Slonegg) [inline gan/trainer.py]: can't you import it in the beginning of the file?

File: apps/scores_plot.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    transform = ResizeTransform()
+    data = Data(conf.DATA_PATH, transform)
+    train_size = int(0.8 * len(data))
+    valid_size = len(data) - train_size
+    train_data, valid_data = torch.utils.data.random_split(data, [train_size, valid_size])
+    train_loader = DataLoader(train_data, batch_size=conf.BATCH_SIZE, num_workers=conf.NUM_WORKERS, shuffle=True)
+    valid_loader = DataLoader(valid_data, batch_size=conf.BATCH_SIZE, num_workers=conf.NUM_WORKERS, shuffle=True)
+    print('Dataset size: ', len(data))
+    if hasattr(conf, 'conditions'):
+        from gan.conditional_gan import Generator5Net, Discriminator5
```
Comment (Slonegg) [inline apps/scores_plot.py:29]: don't really understand all this imports in the code

**Review by Slonegg — APPROVED** [review]

File: performance/estimator.py (hunk 2 above)
Comment (trifonova-n) [reply performance/estimator.py:71]: X1 and X2 can be very big numpy arrays.
X1 += 1 and
X1 /= 2
would modify array inplace
and X1 = (X1 + 1) / 2 would create new array

File: inpainting/visualizer.py (hunk 6 above)
Comment (trifonova-n) [reply inpainting/visualizer.py:49]: x and y are scalars

File: apps/scores_plot.py (hunk 9 above)
Comment (trifonova-n) [reply apps/scores_plot.py:29]: this script can work with both conditional and unconditional gan. But they are defined in different modules
